### PR TITLE
powershell scripts to build and test on windows

### PR DESCRIPTION
**Pull request by sroylance**

### CODE REVIEW

**Comment by sroylance** [comment]
it would be great if at least one other person running go on windows could verify these work

**Comment by chrislovecnm** [comment]
ping @jordansissel see https://github.com/kubernetes/kops/pull/4927#issuecomment-379046668

File: hack/build-windows.ps1 [hunk 1]
```diff
@@ -1,3 +1,21 @@
+#!/bin/bash
```
Comment (chrislovecnm) [inline hack/build-windows.ps1]: umm yah probably do not want that :)

**Comment by jordansissel** [comment]
```
> powershell .\hack\build-windows.ps1


    Directory: D:\gopath\src\k8s.io\kops\.build


Mode                LastWriteTime         Length Name
----                -------------         ------ ----
d-----         4/6/2018   1:02 PM                dist
go-bindata : The term 'go-bindata' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the
spelling of the name, or if a path was included, verify that the path is correct and try again.
At D:\build\kops\hack\build-windows.ps1:25 char:1
+ go-bindata -o upup/models/bindata.go -pkg models -ignore="\\.DS_Store ...
+ ~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (go-bindata:String) [], CommandNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException

go-bindata : The term 'go-bindata' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the
spelling of the name, or if a path was included, verify that the path is correct and try again.
At D:\build\kops\hack\build-windows.ps1:26 char:1
+ go-bindata -o federation/model/bindata.go -pkg model -ignore="\\.DS_S ...
+ ~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (go-bindata:String) [], CommandNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException
```

It looks like `go-bindata` is required? I briefly googled for this and found what claims is an abandoned github repository, so I'm not sure how to proceed.

**Comment by jordansissel** [comment]
Resolving the above I did `go get -u github.com/go-bindata/go-bindata/...` hoping this was the right solution. I also had to add $GOPATH\bin to my $PATH env.

Next, I needed to have `kops` checked out in a different place. $GOPATH/src/k8s.io/kops (forgive my relative comfort with being permanently confused by $GOPATH, still, after 6 years of using Go).

✅ I am successful. :)

```
> powershell .\hack\build-windows.ps1


    Directory: D:\gopath\src\k8s.io\kops\.build


Mode                LastWriteTime         Length Name
----                -------------         ------ ----
d-----         4/6/2018   1:02 PM                dist
```

```
> .\.build\dist\windows\amd64\kops.exe version
Version 0.1.0-winhack (git-6c5e4a753)
```

**Comment by chrislovecnm** [comment]
@jordansissel we have vendored gobindata, and should be installing it from there:

From the makefile:

```
go build ${EXTRA_BUILDFLAGS} ${LDFLAGS}"${EXTRA_LDFLAGS}" -o $@ k8s.io/kops/vendor/github.com/jteeuwen/go-bindata/go-bindata
```

File: hack/build-windows.ps1 [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+$VERSION = '0.1.0-winhack'
+$GOPATH = (go env -json | ConvertFrom-Json).GOPATH
+$BUILD = $GOPATH + '\src\k8s.io\kops\.build'
+$DIST = $BUILD + "\dist"
+New-Item -Force -ItemType Directory -Path $DIST
+$GITSHA = Invoke-Command {cd "$GOPATH/src/k8s.io/kops"; git describe --always }
+$env:GOOS='windows'
+$Env:GOARCH='amd64'
+go-bindata -o upup/models/bindata.go -pkg models -ignore="\\.DS_Store" -ignore="bindata\\.go" -ignore="vfs\\.go" -prefix upup/models/ upup/models/...
```
Comment (jordansissel) [inline hack/build-windows.ps1]: Should this build `go-bindata` first before trying to invoke it?

**Comment by jordansissel** [comment]
> we have vendored gobindata, and should be installing it from there

@chrislovecnm Ahh that's helpful. I added an inline comment in the build-windows.ps1 to ask if about go-bindata building.

**Comment by justinsb** [comment]
I think we should with building from Linux, but cross-building for Windows.  We need the version of kops to be set, which is currently maintained in the makefile.  Building on bazel should work nicely (and avoids having to duplicate the logic around the correct docker version & correct version tags a third time), but is currently broken on windows: https://github.com/bazelbuild/bazel/issues/4496

I propose punting this (easy native building on windows) to 1.11; hopefully we can get bazel working, but if we want to do the Powershell approach we have to get the current version from somewhere.  Likely we should just move it into a standalone file that the Makefile, bazel, and the windows build tooling would all read.

**Comment by justinsb** [comment]
PS this is awesome, it just doesn't really fit the timelines for 1.10 - which is my fault for being slow to get a windows machine up

**Comment by jordansissel** [comment]
This proposal sounds good to me :)

**Comment by sroylance** [comment]
My intention with these scripts was mostly to document the steps to build and test, so others who want to hack kops on windows don't have to figure them out.  Getting make or bazel to work reasonably on windows doesn't seem to be possible right now.  I agree that leaving the official build chain linux-only makes sense until the underlying tools get better.

**Pull request closed without merging**